README
NAME
    CGI::Cookie::XS - HTTP Cookie parser in pure C

VERSION
    This document describes CGI::Cookie::XS 0.18 released on September 2,
    2009.

SYNOPSIS
        use CGI::Cookie::XS;

        my $raw_cookie = 'foo=a%20phrase;weird; bar=yes%2C%20a%20phrase; baz=%5Ewibble&leiyh; qux=%27';
        my $res = CGI::Cookie::XS->parse($raw_cookie);
        # $res is something like:
        #    {
        #      'bar' => [
        #                 'yes, a phrase'
        #               ],
        #      'baz' => [
        #                 '^wibble',
        #                 'leiyh'
        #               ],
        #      'foo' => [
        #                 'a phrase'
        #               ],
        #      'qux' => [
        #                 '\''
        #               ]
        #    };

        # or directly read raw cookies from the CGI environments:
        $res = CGI::Cookie::XS->fetch;

DESCRIPTION
    This module implements a very simple parser for cookies used in HTTP
    applications. We've found CGI::Simple::Cookie and CGI::Cookie rather
    slow according to the profiling results for our OpenResty project, hence
    the rewrite in C.

    This library is still in beta stage and the API is still in flux. We're
    just following the "release early, releaes often" guideline. So please
    check back often ;)

    Special effort has been made to ensure this module works in the same way
    as the latest CGI::Cookie (i.e., the pure Perl implementation). If you
    find it doesn't, please let us know.

METHODS
    We currently provide 2 static methods, "parse" and "fetch". They work
    mostly the same way as those methods found in CGI::Cookie and
    CGI::Simple::Cookie but with the exception that our version returns
    plain Perl data structures rather than hashes of Perl objects (due to
    performance considerations).

    We'll implement some cookie dump methods in the near future.

    "$ref = CGI::Cookie::XS->parse($raw_cookie)"
        Parses $raw_cookie and returns the reference of a hash of arrays.
        The keys of the hash are cookie variables' names while the values of
        the hash are lists of cookie variable's values.

        There is a length limit on the $raw_cookie. If $raw_cookie is longer
        than 4 KB (i.e. 4 * 1024 bytes, excluding the trailing '\0'), the
        overflowing part will be truncated.

        Also note that, "fetch" does not assume any encoding on the cookie
        values. It just decodes the encoded entries verbatim and treat them
        as plain "binary" stuff.

    "$ref = CGI::Cookie::XS->fetch()"
        Reads the raw cookie from the "HTTP_COOKIE" and "COOKIE"
        environments (which are usually set by HTTP servers like lighttd or
        apache) and then parses the value using the "parse" method and
        finally returns the results.

TODO
    *   Removing trailing spaces in cookie values.

SOURCE CONTROL
    For the very latest version of this module, check out the source from
    the Git repos below:

    <http://github.com/agentzh/cookiexs/>

    There is anonymous access to all. If you'd like a commit bit, please let
    us know. :)

BUGS
    There must be some serious bugs lurking somewhere. We haven't done
    comprehensive testing for our code yet. It's a TODO.

    Please report bugs or send wish-list to
    <http://rt.cpan.org/NoAuth/Bugs.html?Dist=CGI-Cookie-XS>.

SEE ALSO
    CGI::Cookie, CGI::Cookie::Simple.

AUTHOR
    yuting "<yuting at alibaba-inc.com>"
    agentzh "<agentzh at yahoo.cn>"

COPYRIGHT
    Copyright (c) 2008 by Yahoo! China EEEE Works, Alibaba Inc. Copyright
    (c) 2009 by Taobao Inc., Alibaba Group.

License
    The "MIT" License

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
0.17 Sep 2 2009
* fixed a parser bug when variables take empty values like "foo=",
  reported and fixed by Colin Keith (rt.cpan.org #49302)

0.16 Sep 12 2008
* Fixed an issue in POD.

0.15 Sep 11 2008
* Some minor optimization (";" is never actually used to separate multiple values for a single key).
* Added Filter::Util::Call to the dependency list.

0.14 Sep 10 2008
* Carefully reviewed the old implementation and gave it a massive rewrite.
* Fixed the issue in #39120
* Fixed a lot of incompatibilities with the latest CGI::Cookie.

0.13 Arg 4 2008
* Fixed the module name in the test suite; we now use CGI::Cookie::XS, rather than Cookie::XS.

0.12 Arg 4 2008
* Fixed a typo in the POD.

0.11 Aug 4 2008
* Renamed the module to CGI::Cookie::XS.

0.10 Aug 4 2008
* Put a notice saying that Cookie::XS is deprecated; one should use CGI::Cookie::XS instead.

0.09 Mar 24 2008
* Enabled the tests in 99-pod-coverage.t.

0.08 Mar 20 2008
* Fixed the foo=ba=r bug reported by RT #34238.
* Fixed a stack overflow when input cookies are too big.
* Added a lot of stuff to the POD documentation.
* Cleaned up the XS code a bit.

0.07 Mar 6 2008
* Fixed "const char*" in XS.xs to make it parsable by perl 5.6.x

0.06 Mar 3 2008
* Fixed the duplicate BUGS section in the POD.
* Fixed the version number in META.yml.

0.05 Mar 3 2008
* Fixed a lot of typos in the POD.

0.04 Mar 3 2008
* initial CPAN release



util.c
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include "INLINE.h"

char** XS_unpack_charPtrPtr(SV* arg) {
  AV* avref;
  char** array;
  STRLEN len;
  SV** elem;
  int i;

  if(!SvROK(arg))
    croak("XS_unpack_charPtrPtr: arg is not a reference");
  if( SvTYPE(SvRV(arg)) != SVt_PVAV)
    croak("XS_unpack_charPtrPtr: arg is not an array");
  avref = (AV*)SvRV(arg);
  len = av_len( avref) + 1;
  array = (char **) SvPVX( sv_2mortal( NEWSV(0, (len +1) * sizeof( char*) )));
  for(i = 0; i < len; i++ ) {
    elem = av_fetch( avref, i, 0);
    array[i] = (char *) SvPV( *elem, PL_na);
  }
  array[len] = NULL;
  return array;
}


void XS_pack_charPtrPtr( SV* arg, char** array, int count) {
  int i;
  AV* avref;

  avref = (AV*) sv_2mortal((SV*) newAV() );
  for( i = 0; i < count; i++) {
    av_push(avref, newSVpv(array[i], strlen(array[i])) );
  }
  SvSetSV( arg, newRV((SV*) avref) );
}



obsolete/inc/Module/Install/Compiler.pm
#line 1
package Module::Install::Compiler;

use strict;
use Module::Install::Base;
use File::Basename ();

use vars qw{$VERSION $ISCORE @ISA};
BEGIN {
	$VERSION = '0.75';
	$ISCORE  = 1;
	@ISA     = qw{Module::Install::Base};
}

sub cc_files {
    require Config;
    my $self = shift;
    $self->makemaker_args(
        OBJECT => join ' ', map { substr($_, 0, -2) . $Config::Config{_o} } @_
    );
}

sub cc_inc_paths {
    my $self = shift;
    $self->makemaker_args(
        INC => join ' ', map { "-I$_" } @_
    );
}

sub cc_lib_paths {
    my $self = shift;
    $self->makemaker_args(
        LIBS => join ' ', map { "-L$_" } @_
    );
}

sub cc_lib_links {
    my $self = shift;
    $self->makemaker_args(
        LIBS => join ' ', $self->makemaker_args->{LIBS}, map { "-l$_" } @_
    );
}

sub cc_optimize_flags {
    my $self = shift;
    $self->makemaker_args(
        OPTIMIZE => join ' ', @_
    );
}

1;

__END__

#line 110


lib/CGI/Cookie/XS.pm
package CGI::Cookie::XS;

use strict;
use warnings;

our $VERSION;

use XSLoader;
BEGIN {
    $VERSION = '0.18';
    XSLoader::load(__PACKAGE__, $VERSION);
}

sub fetch {
    my $class = shift;
    my $raw_cookie = $ENV{HTTP_COOKIE} || $ENV{COOKIE} or return;
    $class->parse($raw_cookie);
}

sub parse {
    _parse_cookie($_[1]);
}

1;
__END__

=head1 NAME

CGI::Cookie::XS - HTTP Cookie parser in pure C

=head1 VERSION

This document describes CGI::Cookie::XS 0.18 released on September 2, 2009.

=head1 SYNOPSIS

    use CGI::Cookie::XS;

    my $raw_cookie = 'foo=a%20phrase;weird; bar=yes%2C%20a%20phrase; baz=%5Ewibble&leiyh; qux=%27';
    my $res = CGI::Cookie::XS->parse($raw_cookie);
    # $res is something like:
    #    {
    #      'bar' => [
    #                 'yes, a phrase'
    #               ],
    #      'baz' => [
    #                 '^wibble',
    #                 'leiyh'
    #               ],
    #      'foo' => [
    #                 'a phrase'
    #               ],
    #      'qux' => [
    #                 '\''
    #               ]
    #    };

    # or directly read raw cookies from the CGI environments:
    $res = CGI::Cookie::XS->fetch;

=head1 DESCRIPTION

This module implements a very simple parser for cookies used in HTTP applications. We've found L<CGI::Simple::Cookie> and L<CGI::Cookie> rather slow according to the profiling results for our L<OpenResty> project, hence the rewrite in C.

Special effort has been made to ensure this module works in the same way as the latest L<CGI::Cookie> (i.e., the pure Perl implementation). If you find it doesn't, please let us know.

=head1 METHODS

We currently provide 2 static methods, C<parse> and C<fetch>. They work mostly the same way as those methods found in L<CGI::Cookie> and L<CGI::Simple::Cookie> but with the exception that our version returns plain Perl data structures rather than hashes of Perl objects (due to performance considerations).

We'll implement some cookie dump methods in the near future.

=over

=item C<< $ref = CGI::Cookie::XS->parse($raw_cookie) >>

Parses C<$raw_cookie> and returns the reference of a hash of arrays. The keys
of the hash are cookie variables' names while the values of the hash are lists of cookie variable's values.

There is a length limit on the C<$raw_cookie>. If C<$raw_cookie> is longer than 4 KB (i.e. 4 * 1024 bytes, excluding the trailing '\0'), the overflowing part will be truncated.

Also note that, C<fetch> does not assume any encoding on the cookie values. It just decodes the encoded entries verbatim and treat them as plain "binary" stuff.

=item C<< $ref = CGI::Cookie::XS->fetch() >>

Reads the raw cookie from the C<HTTP_COOKIE> and C<COOKIE> environments
(which are usually set by HTTP servers like lighttd or apache) and then
parses the value using the C<parse> method and finally returns the
results.

=back

=head1 TODO

=over

=item *

Removing trailing spaces in cookie values.

=back

=head1 SOURCE CONTROL

For the very latest version of this module, check out the source from
the Git repos below:

L<http://github.com/agentzh/cookiexs/>

There is anonymous access to all. If you'd like a commit bit, please let
us know. :)

=head1 BUGS

There must be some serious bugs lurking somewhere. We haven't done comprehensive testing for our code yet. It's a TODO.

Please report bugs or send wish-list to
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=CGI-Cookie-XS>.

=head1 SEE ALSO

L<CGI::Cookie>, L<CGI::Cookie::Simple>.

=head1 AUTHOR

=over

=item yuting C<< <yuting at alibaba-inc.com> >>

=item agentzh C<< <agentzh at yahoo.cn> >>

=back

=head1 COPYRIGHT

Copyright (c) 2008 by Yahoo! China EEEE Works, Alibaba Inc.

Copyright (c) 2009 by Taobao Inc., Alibaba Group.

=head1 License

The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

